File: GD32驱动器/GD32FirmwareUpdater/src/GD32FirmwareUpdater/Protocol/ProtocolFrame.cs
// GD32低压驱动器Boot升级协议 - 协议帧构造与解析
// 负责协议帧的组装和拆解
//
// 帧格式：
// | SOF(1B) | FUNC(1B) | CMD(2B,大端) | DATALEN(2B,小端) | DATA(NB) | CRC(2B,小端) |
// CRC校验范围：从SOF到DATA的所有字节

namespace GD32FirmwareUpdater.Protocol
{
    /// <summary>
    /// 协议帧数据结构
    /// </summary>
    public class ProtocolFrame
    {
        // 帧头（固定0x01）
        public byte Sof { get; set; } = FrameConstants.SOF;

        // 功能码（固定0x06）
        public byte Func { get; set; } = FrameConstants.FUNC;

        // 指令码（大端模式存储）
        public ushort Cmd { get; set; }

        // 有效数据
        public byte[] Data { get; set; } = Array.Empty<byte>();

        // CRC校验值
        public ushort Crc { get; set; }

        /// <summary>
        /// 将协议帧序列化为字节数组（用于发送）
        /// 组帧流程：
        /// 1. 写入SOF + FUNC
        /// 2. 写入CMD（大端：高字节在前）
        /// 3. 写入DATALEN（小端：低字节在前）
        /// 4. 写入DATA
        /// 5. 计算CRC并写入（小端）
        /// </summary>
        public byte[] Serialize()
        {
            int dataLen = Data?.Length ?? 0;
            // 总帧长 = 帧头(6) + 数据(N) + CRC(2)
            int frameLen = FrameConstants.HEADER_SIZE + dataLen + FrameConstants.CRC_SIZE;
            byte[] buffer = new byte[frameLen];

            int offset = 0;

            // SOF
            buffer[offset++] = Sof;

            // FUNC
            buffer[offset++] = Func;

            // CMD - 大端模式（高字节在前）
            buffer[offset++] = (byte)((Cmd >> 8) & 0xFF);
            buffer[offset++] = (byte)(Cmd & 0xFF);

            // DATALEN - 小端模式（低字节在前）
            ushort dataLength = (ushort)dataLen;
            buffer[offset++] = (byte)(dataLength & 0xFF);
            buffer[offset++] = (byte)((dataLength >> 8) & 0xFF);

            // DATA
            if (dataLen > 0)
            {
                Array.Copy(Data!, 0, buffer, offset, dataLen);
                offset += dataLen;
            }

            // CRC - 对SOF到DATA区域计算CRC16，小端写入
            ushort crc = CrcCalculator.Calculate(buffer, 0, offset);
            Crc = crc;
            byte[] crcBytes = CrcCalculator.ToLittleEndianBytes(crc);
            buffer[offset++] = crcBytes[0];
            buffer[offset++] = crcBytes[1];

            return buffer;
        }

        /// <summary>
        /// 从字节数组中反序列化协议帧（用于接收解析）
        /// 解帧流程：
        /// 1. 校验最小长度
        /// 2. 校验SOF和FUNC
        /// 3. 解析CMD（大端）和DATALEN（小端）
        /// 4. 提取DATA
        /// 5. 校验CRC
        /// </summary>
        /// <returns>解析成功返回ProtocolFrame对象，失败返回null</returns>
        public static ProtocolFrame? Deserialize(byte[] buffer, int offset, int length)
        {
            // 最小帧长 = 帧头(6) + CRC(2) = 8字节
            if (length < FrameConstants.HEADER_SIZE + FrameConstants.CRC_SIZE)
                return null;

            int pos = offset;

            // 校验SOF
            byte sof = buffer[pos++];
            if (sof != FrameConstants.SOF)
                return null;

            // 校验FUNC
            byte func = buffer[pos++];
            if (func != FrameConstants.FUNC)
                return null;

            // 解析CMD（大端）
            ushort cmd = (ushort)((buffer[pos] << 8) | buffer[pos + 1]);
            pos += 2;

            // 解析DATALEN（小端）
            ushort dataLen = (ushort)(buffer[pos] | (buffer[pos + 1] << 8));
            pos += 2;

            // 校验总长度是否匹配
            int expectedLen = FrameConstants.HEADER_SIZE + dataLen + FrameConstants.CRC_SIZE;
            if (length < expectedLen)
                return null;

            // 提取DATA
            byte[] data = new byte[dataLen];
            if (dataLen > 0)
            {
                Array.Copy(buffer, pos, data, 0, dataLen);
                pos += dataLen;
            }

            // 提取CRC（小端）
            ushort receivedCrc = (ushort)(buffer[pos] | (buffer[pos + 1] << 8));

            // 计算并校验CRC（从SOF到DATA结束）
            int crcDataLen = FrameConstants.HEADER_SIZE + dataLen;
            ushort calculatedCrc = CrcCalculator.Calculate(buffer, offset, crcDataLen);

            if (receivedCrc != calculatedCrc)
                return null;

            return new ProtocolFrame
            {
                Sof = sof,
                Func = func,
                Cmd = cmd,
                Data = data,
                Crc = receivedCrc
            };
        }

        /// <summary>
        /// 从完整字节数组反序列化
        /// </summary>
        public static ProtocolFrame? Deserialize(byte[] buffer)
        {
            return Deserialize(buffer, 0, buffer.Length);
        }
    }

    /// <summary>
    /// 协议帧构造工厂 - 快速创建各种指令帧
    /// </summary>
    public static class FrameBuilder
    {
        /// <summary>
        /// 构建升级请求帧（CMD=0x0010）
        /// </summary>
        /// <param name="requestType">请求类型：0x01=升级请求, 0x02=查询状态</param>
        public static byte[] BuildUpgradeRequest(byte requestType)
        {
            var frame = new ProtocolFrame
            {
                Cmd = CommandCode.UPGRADE_REQ,
                Data = new byte[] { requestType, 0x00 }  // DATA[1]保留为0
            };
            return frame.Serialize();
        }

        /// <summary>
        /// 构建启动升级帧（CMD=0x0011）
        /// </summary>
        /// <param name="fileSize">固件文件总大小（字节）</param>
        public static byte[] BuildUpgradeStart(uint fileSize)
        {
            var frame = new ProtocolFrame
            {
                Cmd = CommandCode.UPGRADE_START,
                Data = new byte[]
                {
                    (byte)(fileSize & 0xFF),
                    (byte)((fileSize >> 8) & 0xFF),
                    (byte)((fileSize >> 16) & 0xFF),
                    (byte)((fileSize >> 24) & 0xFF)
                }
            };
            return frame.Serialize();
        }

        /// <summary>
        /// 构建发送升级数据帧（CMD=0x0012）
        /// </summary>
        /// <param name="offsetAddr">当前数据在固件文件中的偏移地址</param>
        /// <param name="data">升级数据（不超过128字节）</param>
        public static byte[] BuildUpgradeData(uint offsetAddr, byte[] data)
        {
            // 构建DATA区域：4字节偏移地址 + N字节升级数据
            byte[] frameData = new byte[4 + data.Length];

            // 偏移地址（4字节）
            frameData[0] = (byte)(offsetAddr & 0xFF);
            frameData[1] = (byte)((offsetAddr >> 8) & 0xFF);
            frameData[2] = (byte)((offsetAddr >> 16) & 0xFF);
            frameData[3] = (byte)((offsetAddr >> 24) & 0xFF);

            // 升级数据
            Array.Copy(data, 0, frameData, 4, data.Length);

            var frame = new ProtocolFrame
            {
                Cmd = CommandCode.UPGRADE_DATA,
                Data = frameData
            };
            return frame.Serialize();
        }

        /// <summary>
        /// 构建完成升级帧（CMD=0x0013）
        /// </summary>
        public static byte[] BuildJumpApp()
        {
            var frame = new ProtocolFrame
            {
                Cmd = CommandCode.JUMP_APP,
                Data = new byte[] { 0x01 }  // 0x01=升级完成请求
            };
            return frame.Serialize();
        }

        /// <summary>
        /// 构建版本号查询帧（CMD=0x0014）
        /// </summary>
        public static byte[] BuildVersionRequest()
        {
            var frame = new ProtocolFrame
            {
                Cmd = CommandCode.VER_REQ,
                Data = new byte[] { 0x01 }  // 0x01=请求版本号
            };
            return frame.Serialize();
        }
    }
}

File: GD32驱动器/GD32FirmwareUpdaterQt/src/protocol/ProtocolFrame.cpp
// ============================================================
//  协议帧序列化 / 反序列化实现
//
//  【Qt知识点】大小端：
//  - 大端（Big-Endian）：高位字节在前，用于 CMD
//  - 小端（Little-Endian）：低位字节在前，用于 DATALEN 和 CRC
//  注意与 WPF 版完全一致，保证串口线上字节流逐字节兼容。
// ============================================================
#include "ProtocolFrame.h"
#include "CrcCalculator.h"

// ==================== 序列化（发送方向） ====================
QByteArray ProtocolFrame::serialize() const
{
    QByteArray buffer;

    // 1. 帧头 + 功能码
    buffer.append(static_cast<char>(sof));
    buffer.append(static_cast<char>(func));

    // 2. CMD — 大端（高字节在前）
    buffer.append(static_cast<char>((cmd >> 8) & 0xFF));
    buffer.append(static_cast<char>(cmd & 0xFF));

    // 3. DATALEN — 小端（低字节在前）
    quint16 dataLen = static_cast<quint16>(data.size());
    buffer.append(static_cast<char>(dataLen & 0xFF));
    buffer.append(static_cast<char>((dataLen >> 8) & 0xFF));

    // 4. DATA
    buffer.append(data);

    // 5. CRC — 对 SOF 到 DATA 区域计算，小端写入
    quint16 crcValue = CrcCalculator::calculate(buffer);
    buffer.append(CrcCalculator::toLittleEndianBytes(crcValue));

    return buffer;
}

// ==================== 反序列化（接收方向） ====================
ProtocolFrame ProtocolFrame::deserialize(const QByteArray &buffer, bool *ok)
{
    ProtocolFrame frame;

    // 最小帧长 = 帧头(6) + CRC(2) = 8 字节
    if (buffer.size() < FrameConstants::HEADER_SIZE + FrameConstants::CRC_SIZE)
    {
        if (ok) *ok = false;
        return frame;
    }

    // 帧头字段
    frame.sof  = static_cast<quint8>(buffer.at(0));
    frame.func = static_cast<quint8>(buffer.at(1));

    // 解析 CMD（大端）
    frame.cmd = (static_cast<quint16>(static_cast<quint8>(buffer.at(2))) << 8)
              |  static_cast<quint8>(buffer.at(3));

    // 解析 DATALEN（小端）
    quint16 dataLen = static_cast<quint8>(buffer.at(4))
                    | (static_cast<quint16>(static_cast<quint8>(buffer.at(5))) << 8);

    // 总帧长校验
    int frameLen = FrameConstants::HEADER_SIZE + dataLen + FrameConstants::CRC_SIZE;
    if (buffer.size() < frameLen)
    {
        if (ok) *ok = false;
        return frame;
    }

    // 提取 DATA
    frame.data = buffer.mid(FrameConstants::HEADER_SIZE, dataLen);

    // 提取 CRC（小端）
    int crcPos = FrameConstants::HEADER_SIZE + dataLen;
    frame.crc = static_cast<quint8>(buffer.at(crcPos))
              | (static_cast<quint16>(static_cast<quint8>(buffer.at(crcPos + 1))) << 8);

    // 校验帧头固定值
    if (frame.sof != FrameConstants::SOF || frame.func != FrameConstants::FUNC)
    {
        if (ok) *ok = false;
        return frame;
    }

    // 计算并比对 CRC（范围：SOF 到 DATA 结束）
    quint16 calcCrc = CrcCalculator::calculate(buffer, 0, FrameConstants::HEADER_SIZE + dataLen);
    if (frame.crc != calcCrc)
    {
        if (ok) *ok = false;
        return frame;
    }

    if (ok) *ok = true;
    return frame;
}

// ==================== 帧构造工厂实现 ====================

QByteArray FrameBuilder::buildUpgradeRequest(quint8 requestType)
{
    // DATA = [请求类型, 保留0x00]
    QByteArray data;
    data.append(static_cast<char>(requestType));
    data.append(static_cast<char>(0x00));

    ProtocolFrame frame;
    frame.cmd = CommandCode::UPGRADE_REQ;
    frame.data = data;
    return frame.serialize();
}

QByteArray FrameBuilder::buildUpgradeStart(quint32 fileSize)
{
    // DATA = 文件总大小（4 字节，小端）
    QByteArray data;
    data.append(static_cast<char>(fileSize & 0xFF));
    data.append(static_cast<char>((fileSize >> 8) & 0xFF));
    data.append(static_cast<char>((fileSize >> 16) & 0xFF));
    data.append(static_cast<char>((fileSize >> 24) & 0xFF));

    ProtocolFrame frame;
    frame.cmd = CommandCode::UPGRADE_START;
    frame.data = data;
    return frame.serialize();
}

QByteArray FrameBuilder::buildUpgradeData(quint32 offsetAddr, const QByteArray &data)
{
    // DATA = [4字节偏移地址(小端)] + [升级数据]
    QByteArray frameData;
    frameData.append(static_cast<char>(offsetAddr & 0xFF));
    frameData.append(static_cast<char>((offsetAddr >> 8) & 0xFF));
    frameData.append(static_cast<char>((offsetAddr >> 16) & 0xFF));
    frameData.append(static_cast<char>((offsetAddr >> 24) & 0xFF));
    frameData.append(data);

    ProtocolFrame frame;
    frame.cmd = CommandCode::UPGRADE_DATA;
    frame.data = frameData;
    return frame.serialize();
}

QByteArray FrameBuilder::buildJumpApp()
{
    // DATA = [0x01 = 升级完成请求]
    QByteArray data;
    data.append(static_cast<char>(0x01));

    ProtocolFrame frame;
    frame.cmd = CommandCode::JUMP_APP;
    frame.data = data;
    return frame.serialize();
}

QByteArray FrameBuilder::buildVersionRequest()
{
    // DATA = [0x01 = 请求版本号]
    QByteArray data;
    data.append(static_cast<char>(0x01));

    ProtocolFrame frame;
    frame.cmd = CommandCode::VER_REQ;
    frame.data = data;
    return frame.serialize();
}
